knimin/lib/data_access.py

```diff
@@ -3894,6 +3894,94 @@ def delete_shotgun_index_aliquot(self, shotgun_index_aliquot_id):
+            each of the 384 unique i7 and i5 barcodes 1-384 (going across
+            rows), in initial plate of 384 samples sample 1 will get i5_1
+            and i7_1, sample 2 will get i5_2 and i7_2, and so on. For the
+            next plate of 384 samples, sample 1 will get i5_1 and i7_2,
+            sample 2 will get i5_2 and i7_3, and so on. After 147,456
+            samples it will reset back to i5_1 and i7_1.
+        """
+        with TRN:
+            sql = """SELECT shotgun_index_tech_id, name, dual_index,
+                        last_index_idx
+                     FROM pm.shotgun_index_tech WHERE name = %s"""
+            TRN.add(sql, [idx_tech])
+            idx_tech_vals = TRN.execute_fetchindex()
+
+            if not idx_tech_vals:
+                raise ValueError('Not a valid shotgun index '
+                                 'technology: %s' % idx_tech)
+            else:
+                # just taking the first element and converting it to dict
+                idx_tech_vals = dict([v for v in idx_tech_vals][0])
+
+            # selecting single indexing
+            sql = """SELECT shotgun_index_id
+                     FROM pm.shotgun_index
+                     WHERE shotgun_index_tech_id = %s AND index_type = 'i5'
+                     ORDER BY shotgun_index_id"""
+            TRN.add(sql, [idx_tech_vals['shotgun_index_tech_id']])
+            indices = [r[0] for r in TRN.execute_fetchindex()]
+            if not indices:
+                raise ValueError("%s doesn't have any i5 values" % idx_tech)
+
+            if idx_tech_vals['dual_index']:
+                # selecting the dual section
+                sql = """SELECT shotgun_index_id
+                         FROM pm.shotgun_index
+                         WHERE shotgun_index_tech_id = %s AND index_type = 'i7'
+                         ORDER BY shotgun_index_id"""
+                TRN.add(sql, [idx_tech_vals['shotgun_index_tech_id']])
+                indices_i7 = [r[0] for r in TRN.execute_fetchindex()]
+                if not indices_i7:
+                    raise ValueError("%s doesn't have any i7 "
+                                     "values" % idx_tech)
+
+                indices = list(product(indices, indices_i7))
+
+            indices_generated = []
+            len_indices = len(indices)
+            # note that this assumes that we will never generate more samples
+            # in a single go than avialable indices
+            idx_start = idx_tech_vals['last_index_idx']
+            while len(indices_generated) < num_samples:
+                idx_end = idx_start + num_samples
+
+                if idx_end > len_indices:
+                    idx_end = len_indices
+                indices_generated.extend(indices[idx_start:idx_end])
+
+                # in case we have another loop we need to start in 0 and
+                # reset the num_samples requested
+                idx_start = 0
```

Comment:
`idx_start` will always be zero with the exception of the first time through the loop, is that correct?